ddtrace/internal/writer/writer.py

```diff
@@ -420,14 +420,17 @@ def _flush_queue_with_client(self, client, raise_exc=False):
                 )
         finally:
             if config.health_metrics_enabled and self.dogstatsd:
+                # _metrics is updated when traces are queued.
+                # Here we get a copy of the metrics dictionary as early as possible
+                metrics = self._metrics.copy()
```

Comment:
Is copying better than guarding this data structure with a lock, in terms of performance? (I'm assuming the issue occurs because of concurrent access from different threads here).

Reply:
We add values to the metrics dictionary multiple times each times per trace chunk (a few times when `tracer.flush()` is called and a few times when `tracer.write()` is called). Acquiring a lock every time `_metrics_dist` is invoked could come with a significant performance hit.

The current approach hides copying the dictionary behind a the `health_metrics_enabled` flag. This flag is disabled by default so most users will not get this performance. We also need a deep copy here since self._metrics has nested dictionaries. I made this fix in my latest commit.

Reply:
But `flush` would be called very rarely compared to `write`, so most of the time the lock would be acquired immediately. I have no feel for how big the metrics dict could be, so I wonder if the amortised cost of deepcopy would really be cheaper than locking 🤔

Reply:
Are we, in effect, implementing a thread-safe batching for dogstatsd (which at the time when this code was added had an open issue)? 

This issue looks to be resolved: https://github.com/DataDog/datadogpy/issues/439

We should remove the metrics batching currently implemented in the `AgentWriter`.

Reply:
So instead of batching metrics we should call `self.dogstatsd.distribution(....)` directly?

Reply:
@mabdinur right

Reply:
cc @Kyle-Verhoog and @brettlangdon who can also speak to the choice to batch metrics within the `AgentWriter`.

Reply:
> We add values to the metrics dictionary multiple times each times per trace chunk (a few times when tracer.flush() is called and a few times when tracer.write() is called). Acquiring a lock every time _metrics_dist is invoked could come with a significant performance hit.

I think this is correct, iirc. Performance improvement, take the big hit once at time we flush metrics, rather than a lock every time we create them.

Reply:
I updated the PR description. `TracerWriter._metrics` is no longer used to batch health metrics. It is now only used to calculate `tracer_kr` (removing `TracerWriter._metrics` will require a significant refactor to how we calculate `tracer_kr`).